### ISSUE TITLE
Non descriptive error message when setting up SendOnly endpoint without persistence.

### ISSUE THREAD

**Issue by dvdstelt** [Particular/NServiceBus#3748]
The following error occurs when I set up a SendOnly endpoint without persistence.

```
An unhandled exception of type 'System.Collections.Generic.KeyNotFoundException' occurred in mscorlib.dll

Additional information: The given key (ResultingSupportedStorages) was not present in the dictionary.
```

Commenting out the persistence line fixes it.

```
var config = new EndpointConfiguration("Website.Application");
config.SendOnly();

config.UseTransport<MsmqTransport>().ConnectionString("deadLetter=false;journal=false");
config.UseSerialization<JsonSerializer>();
//config.UsePersistence<InMemoryPersistence>();

config.SendFailedMessagesTo("error");

ConventionsBuilder conventions = config.Conventions();
conventions.DefiningCommandsAs(t => t.Namespace != null && t.Namespace == "Divergent.Messages" || t.Name.EndsWith("Command"));
conventions.DefiningEventsAs(t => t.Namespace != null && t.Namespace == "Divergent.Messages" || t.Name.EndsWith("Event"));

var endpoint = Endpoint.Start(config).GetAwaiter().GetResult();
```

From a user perspective I would say it should not look for storage, but the error message could also be more descriptive.

**Comment by bording** [comment]
This is an interesting problem. The `KeyNotFoundException` is thrown because each feature that requires some sort of persistenece storage now checks to see if the configured persistenece supports that storage type. A SendOnly endpoint can be allowed to not have a persistence at all, so the `MessageDrivenSubscriptions` feature is checking for subscription storage support, and then hitting the missing key.

While the `PersistenceStartup.HasSupportFor()` check could be modified to account for this case, there is a larger issue here:

We have a inconsistency on support for publishing messages from a SendOnly endpoint. Even if we fixed the `KeyNotFoundException`, a SendOnly endpoint doesn't have a queue, so it can never receive subscription messages.

However, for a transport with native pubs/sub support, a SendOnly endpoint can publish messages just fine without needing its own queue.

**Comment by dvdstelt** [comment]
So what you're saying is, it should be renamed to `SendOnlyUnlessTheTransportSupportsNativePubSubThenEndpointAlsoSupportsPublish` :smile: 

But kidding aside, in my opinion the `SendOnly` endpoints do exactly what they're called, they only send. So even having `SendLocal` on that interface/api seems illogical, let alone `Publish`. The main reason I would use a SendOnly endpoint is because I'm using it in from a UI and/or I want to have an extremely short startup time. Perhaps in some batch process where I send a lot of messages. But it's **send only** so it's not logical to have to configure persistence.

Now I get why we use callbacks, because of backwards compatibility and because clients find it easy to move to NServiceBus in an RPC world with a legacy system. Perhaps there are reasons that we support `Publish` on a `SendOnly` endpoint. But putting those possible reasons aside, I can't think of a reason why we'd want to publish on a `SendOnly` endpoint, nor configure persistence.

**Comment by danielmarbach** [comment]
Bundled it with https://github.com/Particular/PlatformDevelopment/issues/734

**Comment by dvdstelt** [comment]
And removed it again!? :)

**Comment by bording** [comment]
Personally, I wouldn't want to interpret the name so literally! While the scenarios for publishing from a web frontend might not be as abundant as sending from one, I think I'd prefer to not completely rule out the possibility, especially since there are transports that support it perfectly fine.

**Comment by dvdstelt** [comment]
> Personally, I wouldn't want to interpret the name so literally!

We made large API changes, like changing from BusConfiguration to EndpointConfiguration, just for the sake of the name. So not taking SendOnly literally seems off.

We abstract away the native pub/sub possibilities (including storing subscriptions) with NServiceBus. It doesn't seem right to enable this on special occasions just because the transport makes it possible. In other words, we don't support it on MSMQ & SQL, but ASB we do support publishing from a SendOnly endpoint. I'm completely unaware of how this does/should work, because SendOnly endpoints don't have an incoming queue for subscriptions? Should users suddenly use native ASB to create subscriptions? A question, because I really don't know how this is handled atm without an incoming queue.

Also, the transports you're talking about, might not implement the best practices we advocate? We have [documentation on why not to publish msgs from a website](http://docs.particular.net/nservicebus/hosting/publishing-from-web-applications). I really feel we should (and have always done so) drive the users towards specific scenarios or design considerations, making it hard(er) to make mistakes. This is a feature of NServiceBus that a lot of developers like and is one of the things that makes NServiceBus better than others.

**Comment by bording** [comment]
> We made large API changes, like changing from BusConfiguration to EndpointConfiguration, just for the sake of the name. So not taking SendOnly literally seems off.

While I'm guessing the name "SendOnly" is in the "sending vs. receiving" sense, since we also have a "sending vs. publishing" distinction, perhaps renaming would be a good idea. Not sure that would be in scope for V6 at this point, though.

I wasn't aware of that documentation, but given that we do talk about being able to publish there, and how to work around publishing from a SendOnly endpoint when using a message-based pub/sub transport, perhaps just fixing the `KeyNotFoundException` is enough here?

**Comment by bording** [comment]
So, if the `KeyNotFoundException` was prevented and instead in this scenario, you saw the following error message:

"The selected persistence doesn't have support for subscription storage. Select another persistence or disable the message-driven subscriptions feature using endpointConfiguration.DisableFeature<MessageDrivenSubscriptions>()"

Would that be enough, or would that still be confusing since you actually haven't selected a persistence at all?

**Comment by dvdstelt** [comment]
No, I actually love this.

If people see "selected persistence" and start wondering why it says "selected", when they didn't select a persistence, they'll figure out on their own that there's a default persister used. So it is still a selection, but they didn't know they selected. But then they'll start wondering what else is a default setting and hopefully spend some time on how NServiceBus actually works! :smiley: 

To make a long story short, I'd keep the message you're suggesting.

The only side-note could be that we hint towards more EndpointConfiguration settings, we should probably report consistently with (about) the same message. It's a bit late now and had a few drinks, can't remember which message I saw where! :grin:

**Comment by mikeminutillo** [comment]
Alternatively, the transport could be responsible for enabling by default the `Message Driven Subscription` feature which could depend on a `Persistence` feature and a `Receive` feature. If you don't have both then it doesn't get switched on. If you do have both then we can ask the Persistence feature at startup if it supports subscription storage and blow up if not. 

Our feature landscape is remarkably flat given how much infrastructure there is in place to ensure ordering and dependencies and I suspect the historical reliance on `IsSendOnly` is largely to blame.

**Comment by bording** [comment]
> Alternatively, the transport could be responsible for enabling by default the `Message Driven Subscription` feature which could depend on a `Persistence` feature and a `Receive` feature

I spent some time looking at this, and while it's a change that could be made, I'm not sure I see the benefit of inverting the dependency at this point.

Currently, all of the persistence subscription storage features depend on `MessageDrivenSubscriptions`, so disabling that ends up disabling the entire subscription storage registration, which then causes the endpoint to blow up with an `'Autofac.Core.DependencyResolutionException` because there is no registered `ISubscriptionStorage`.

Since we do support publishing from a SendOnly endpoint, disabling this feature when SendOnly isn't what we want to do. Ultimately, we need an error message that lets end users know that they need to configure a persistence in this scenario.

**Comment by timbussmann** [comment]
I think we can close this issue @bording ?

**Comment by bording** [comment]
Yes, fixed by #3793.
